Adding abridge-diff

jdtsmith opened the pull request:
### Brief summary of what the package does

A simple Emacs package for abridging refined diff hunks (for example in magit). 

### Direct link to the package repository

https://github.com/jdtsmith/abridge-diff

### Your association with the package

Author and maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

jdtsmith commented:
> Thanks for this. Quick pass -

Thank _you_!

> ### abridge-diff.el
> byte-compile (using Emacs 27.1):
> 
> ```
> abridge-diff.el:196:1:Warning: the function `transient-append-suffix' is not
>     known to be defined.
> ```

I now `(require 'transient)` (although `magit` does as well, so this compiler error seems spurious). 

> checkdoc (using version 0.6.2):
> 
> ```
> Warning (emacs): 
> abridge-diff.el:151: Argument `rest' should appear (as REST) in the doc string
> ```

Added.

> ### Package
> * GitHub API found `GNU Affero General Public License v3.0`
> * abridge-diff.el (GPL) -- Abridge long line-based diff hunks, including in magit
> 
> Based on the above I believe you have GPL boilerplate in your file, but are using the GNU Affero in your repo's license. MELPA typically prefers GPL if you're on the fence.

Thanks. Switched to GPL2.

riscy commented:
> I now (require 'transient) (although magit does as well, so this compiler error seems spurious).

Taking a closer look, I see the compiler won't actually run the code in the `when` block during compilation, so it won't actually see the `require` at compile time.  You could for example `(eval-when-compile (require 'transient))` but it would probably be even simpler to just `(require 'transient)` at the top level.

Also now that I'm looking more closely, this:

> Automatically configures itself to work with magit, 

is convenient but it's something we try to push back against -- simply loading a package shouldn't alter Emacs's behavior.  Packages that violate this rule can create grief for users.  It means the user (or starter kit) has to be more explicit/do more work, but it's worth it.

What I'd recommend in this case is to define a minor mode something approximately (forgive any errors) like this:

```elisp
;;;###autoload
(define-minor-mode abridge-diff-mode
  "Toggle abridged text hiding."
  :lighter " xyz"
  (if abridge-diff-mode
      (abridge-diff-enable-hiding) ;<-- this function could call transient-append-suffix
    (abridge-diff-disable-hiding)) ; <-- this one could call transient-remove-suffix
```
and then a globalized minor mode to make it automatically turn itself on when appropriate:
```elisp
;;;###autoload
(define-globalized-minor-mode abridge-diff-global-mode abridge-diff-mode abridge-diff--global-on)

(defun abridge-diff--global-on ()
  "..."
  ;; actually you'll want to handle 'magit-diff-mode as well:
  (when (eq major-mode 'magit-status-mode) (abridge-diff-mode 1)))
```
and then you won't have to `(defvar abridge-diff-hiding nil)` (a variable called `abridge-diff-mode` will be defined).

Otherwise this looks very useful -- git and LaTeX have been a difficult combination.

jdtsmith commented:
> > I now (require 'transient) (although magit does as well, so this compiler error seems spurious).
> 
> Taking a closer look, I see the compiler won't actually run the code in the `when` block during compilation, so it won't actually see the `require` at compile time. You could for example `(eval-when-compile (require 'transient))` but it would probably be even simpler to just `(require 'transient)` at the top level.

I'm not sure that works.  I don't want the package to require `magit` or `transient`, since it may be useful for any refined `diff-mode` content (note that `magit` delegates to `diff-mode` to actually do the refining and diff display).  See [this discussion](https://github.com/magit/magit/issues/1615) for more background.  Any suggestions?

> Also now that I'm looking more closely, this:
> 
> > Automatically configures itself to work with magit,
> 
> is convenient but it's something we try to push back against -- simply loading a package shouldn't alter Emacs's behavior. Packages that violate this rule can create grief for users. It means the user (or starter kit) has to be more explicit/do more work, but it's worth it.
> 
> What I'd recommend in this case is to define a minor mode something approximately (forgive any errors) like this:
> 
> ```emacs-lisp
> ;;;###autoload
> (define-minor-mode abridge-diff-mode
>   "Toggle abridged text hiding."
>   :lighter " xyz"
>   (if abridge-diff-mode
>       (abridge-diff-enable-hiding) ;<-- this function could call transient-append-suffix
>     (abridge-diff-disable-hiding)) ; <-- this one could call transient-remove-suffix
> ```
> 
> and then a globalized minor mode to make it automatically turn itself on when appropriate:
> 
> ```emacs-lisp
> ;;;###autoload
> (define-globalized-minor-mode abridge-diff-global-mode abridge-diff-mode abridge-diff--global-on)
> 
> (defun abridge-diff--global-on ()
>   "..."
>   ;; actually you'll want to handle 'magit-diff-mode as well:
>   (when (eq major-mode 'magit-status-mode) (abridge-diff-mode 1)))
> ```
> 
> and then you won't have to `(defvar abridge-diff-hiding nil)` (a variable called `abridge-diff-mode` will be defined).

Hmm, adding a minor mode that introduces a single new transient command into `magit` seems to be overkill.  Some packages add much larger command space than that on load (e.g. [magit-imerge](https://github.com/magit/magit-imerge/blob/master/magit-imerge.el#L24)). If the objection is solely to the addition of the transient command, perhaps I could add a customize variable `abridge-diff-add-magit-command` that is `nil` by default?  Or is the objection to enabling the functionality at all on refined diff-mode output (which is accomplished via advising `smerge-refine-regions`).

> Otherwise this looks very useful -- git and LaTeX have been a difficult combination.

Yes, it was always an issue.  Thanks for the feedback.

riscy commented:
> Any suggestions?

Ah, I see now.  You can pacify the byte compiler in this case with `(declare-function transient-append-suffix "transient.el" t t)` (for example) and if you ever want to double check you've got it right you can use `check-declare-file`.

> Some packages add much larger command space than that

That's a fair point -- I think this is a fairly minor case and some packages don't do it.  But they should, in my opinion :)

I've been thinking about documenting this somewhere and just linking people, but here's the motivation:

- Say I load your package; `(advice-add #'smerge-refine-regions :after #'abridge-diff-abridge)` runs automatically
- I find I don't like the change, or it interferes with my configuration, or it has a bug
- I look for a way to disable the feature.  The idiomatic way in Emacs is to toggle a minor mode off, but I'm out of luck
- That's okay, Emacs has <kbd>M-x unload-feature</kbd>, and since you `(provide 'abridge-diff)`, so you're providing a feature that I can unload
- I unload the feature, but the advice wasn't removed; now whenever `smerge-refine-regions` runs, Emacs throws an error
- Unless I want to start debugging, I have to restart Emacs (and hope I have removed every reference to the package from my config)

 `unload-feature` is less desirable than toggling a minor mode off.  It is not well known, doesn't remove advice (but it _can_ be made to; see the help on `unload-feature`) and can simply fail if the feature is in use by other code.  And if users aren't sure what's broken with their Emacs, it's much simpler for them to try toggling minor modes off and on (especially if it's one they just enabled).  Moreover, if they're trying to debug their init scripts, it's _much_ easier for them to scan for modes being enabled, rather than teasing apart the dependency tree of packages that's being loaded.

Anyway, in this case I wonder if you could wrap the activation of the advice into a (globalized) minor mode.  Once you do that, it feels natural for the hooks and transient-append-suffix to also be done (and undone) as part of a minor mode.

jdtsmith commented:
> > Any suggestions?
> 
> Ah, I see now. You can pacify the byte compiler in this case with `(declare-function transient-append-suffix "transient.el" t t)` (for example) and if you ever want to double check you've got it right you can use `check-declare-file`.

OK done.

> * Say I load your package; `(advice-add #'smerge-refine-regions :after #'abridge-diff-abridge)` runs automatically
> * I find I don't like the change, or it interferes with my configuration, or it has a bug
> * I look for a way to disable the feature.  The idiomatic way in Emacs is to toggle a minor mode off, but I'm out of luck
> * That's okay, Emacs has M-x unload-feature, and since you `(provide 'abridge-diff)`, so you're providing a feature that I can unload
> * I unload the feature, but the advice wasn't removed; now whenever `smerge-refine-regions` runs, Emacs throws an error
> * Unless I want to start debugging, I have to restart Emacs (and hope I have removed every reference to the package from my config)

This is a compelling argument.  OK, I introduced a _global_ minor mode.  It doesn't enable/disable whether abridged text is hidden, but instead just controls the presence of the advice, and the magit/transient command insertion/removal.  The reason for this is that `advice-add` is always global in scope, so it doesn't make sense to enable/disable that buffer-by-buffer, whereas whether to actively hide the text marked for abridging may vary buffer by buffer.  I suppose I could use the presence of a buffer local variable to determine whether to even bother to compute the abridge-able regions, but since magit only does diff refinement once on load (when `magit-diff-refine-hunk` is set to `all`, as recommended), you'd then have to reload the diff to see any updates after you toggle abridging on.  I do now add a toggle status update with a warning regarding magit diff-refinement not being enabled.  

> Anyway, in this case I wonder if you could wrap the activation of the advice into a (globalized) minor mode. Once you do that, it feels natural for the hooks and transient-append-suffix to also be done (and undone) as part of a minor mode.

This is the approach I took, except I made the minor mode global by default.  So you should be able to disable the minor mode and unload the package and have all traces disappear.  

I now worry a bit about load order.  Is there a better test than:

```
  (when (fboundp 'magit)		;add magit command
```

that will work even if `magit` is not yet loaded?  Also, is it canonical MELPA practice for packages implementing global minor modes to enable them by default on load, or instead to suggest and expect users to `(use-package 'foo (foo-minor-mode t))` or similar?

riscy commented:
You're right about the minor mode being more appropriate, sorry about that.

This looks good -- there's only a couple compile warnings now due to the change, easily fixed with:
```elisp
(declare-function transient-remove-suffix "transient.el" t t)
(defvar magit-diff-refine-hunk)
```

jdtsmith commented:
Thanks, done.  Any thoughts on enabling the minor mode "by default" and load order w.r.t magit/transient?

I could just recommend loading with:

```elisp
(use-package abridge-diff
  :after magit
  :init (abridge-diff-mode 1))
```

but not sure how many MELPA package users expect things to "just work" out of the box.

riscy commented:
`magit-diff-refine-hunk` is being referenced earlier in the file than the defvar, so it's still throwing a warning.  Usually I just prefer to put all my defvars and declare-functions at the top of files to avoid this.

> I could just recommend loading

That's definitely the way to do it.  It _does_ force users (or the maintainers of starter packs) to do a little bit more, but if more packages play by these rules then there'll be fewer stories of Emacs users running into incomprehensible breakages.

jdtsmith commented:
> `magit-diff-refine-hunk` is being referenced earlier in the file than the defvar, 

Fixed.

riscy commented:
Great, thanks!
